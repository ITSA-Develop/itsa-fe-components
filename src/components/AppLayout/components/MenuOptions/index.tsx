
import { TExtendedMenuItem, TMenuMode } from "@/types";
import { Menu, MenuProps } from "antd";
import React from "react";

export interface MenuOptionsProps {
	collapsed: boolean;
	currentPath: string;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	items?: TExtendedMenuItem[];
	mode?: TMenuMode;
	openKeys?: string[];
	onOpenKeysChange?: (openKeys: string[]) => void;
}

const MenuOptions: React.FC<MenuOptionsProps> = ({
  collapsed,
  currentPath = "",
  onClickOptionMenu,
  items = [],
  mode = "inline",
  openKeys = [],
  onOpenKeysChange,
}) => {

  const handleMenuClick: MenuProps['onClick'] = (info) => {
    // Find the clicked item in the menu data
    const findMenuItem = (menuItems: TExtendedMenuItem[], key: string): TExtendedMenuItem | null => {
      for (const item of menuItems) {
        if (item.key === key) {
          return item;
        }
        if ('children' in item && item.children) {
          const found = findMenuItem(item.children as TExtendedMenuItem[], key);
          if (found) return found;
        }
      }
      return null;
    };

    const clickedItem = findMenuItem(items, info.key);
    if (clickedItem && onClickOptionMenu) {
      onClickOptionMenu({ key: info.key, item: clickedItem });
    }
  };

  return (
    <div className="menu-options">
      <Menu
        mode={mode}
        inlineCollapsed={collapsed}
        items={items}
        className="overflow-auto scrollbar-none border-none min-w-full"
        defaultSelectedKeys={[currentPath]}
        openKeys={openKeys}
        onOpenChange={onOpenKeysChange}
        onClick={handleMenuClick}
      />
    </div>
  );
};

export default MenuOptions;
