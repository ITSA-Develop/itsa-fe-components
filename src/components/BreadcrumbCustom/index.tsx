import { Button } from 'antd';
import { LeftOutlined } from '@ant-design/icons';

export interface IBreadcrumbCustomProps {
	title: string;
	description: string;
	action: () => void;
}
export const BreadcrumbCustom = ({ title, description, action }: IBreadcrumbCustomProps) => {
	return (
		<div className="flex flex-row items-center gap-2">
			<Button onClick={action} type="text" icon={<LeftOutlined className="font-bold" style={{ fontSize: '20px' }} />} />
			<div className="flex flex-row items-center gap-2">
				<h4 className="font-bold text-lg text-black-100">{title}</h4>
				<span className="text-black-100">|</span>
				<small className="text-sm text-black-100">{description}</small>
			</div>
		</div>
	);
};
