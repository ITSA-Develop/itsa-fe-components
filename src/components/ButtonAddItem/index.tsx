import { Button, ButtonProps } from 'antd';
import { PlusOutlined } from '@ant-design/icons';

interface IButtonAddItemProps extends ButtonProps {
	label: string;
	onClick: () => void;
}

export const ButtonAddItem = ({ label, onClick,  ...props }: IButtonAddItemProps) => {
	return (
		<Button
            className={`h-6 bg-white !border !border-gray-900 text-gray-900 hover:!border-gray-900 hover:!bg-gray-50 hover:!text-gray-900 rounded-lg`}
			onClick={onClick}
			icon={<PlusOutlined className='text-gray-900 text-xs' />}
			{...props}
		>
			{label}
		</Button>
	);
};
