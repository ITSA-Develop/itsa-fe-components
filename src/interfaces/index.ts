export interface IGetPermissionResponse {
	code: number;
	message: string;
	result: Result;
}

export interface Result {
	agencies: IAgency[];
}

export interface IAgency {
	id: number;
	name: string;
	modules: IModule[];
}

export interface IModule {
	id: number;
	name: string;
	path: string | null;
	icon: string;
	entorno: string;
	submodules: ISubmodule[];
}

export interface ISubmodule {
	id: number;
	name: string;
	path: string | null;
	icon: string | null;
	url?: string | null;
	groups?: ISubmodule[];
	programs?: ISubmodule[];
	actions?: IActions;
}

export interface IActions {
	all_actions: number;
	read: number;
	create: number;
	update: number;
	delete: number;
}

export interface ISelectOptionDropdownButton {
	items: {
		key: string;
		label: string;
		onClick?: (id: string) => void;
	}[];
}

export interface IValidateRouteResponse {
	code: number;
	message: string;
	result: boolean;
}
//--------------------------------------------------->
// TODO: clean up this file after implementations starts
import { EAddressType, EEmailType, EPhoneConnectionType, EPhoneType } from '@/enums';
import { TDate, TInputOptions, TInputRules } from '@/types';
import { Dayjs } from 'dayjs';
import { ReactNode } from 'react';

export interface IBaseInputProps {
	name: string;
	label: string;
	placeholder?: string;
	rules?: TInputRules;
	error?: string;
	isDisabled?: boolean;
	defaultValue?: string | boolean | number | TInputOptions | TInputOptions[] | null | undefined;
	maxLen?: number;
}

export interface IBaseInputDate extends IBaseInputProps {
	minDate?: Dayjs;
	maxDate?: Dayjs;
	isDisableFuture?: boolean;
	isDisablePast?: boolean;
	shouldOpen?: boolean;
	onClose?: () => void;
	onOpen?: () => void;
}

// TODO: double check with BE what is going to be the Iid returned by default.
export interface IId {
	id: string;
	modified_by: string;
	created_by: string;
	modified_on: TDate;
	created_on: TDate;
}

export interface IManualAddressObject {
	principalStreet: string;
	latitude: number;
	longitude: number;
	postalCode: string;
	province?: number;
	canton?: number;
	parish?: number;
}

export interface ILocationFormData {
	province: string | TInputOptions | null;
	canton: string | TInputOptions | null;
	parish: string | TInputOptions | null;
}

export interface IMultipleEmailItem {
	emailType: EEmailType | string;
	email: string;
	isPrincipal: boolean;
}

export interface IMultiplePhoneItem {
	phoneConnectionType: EPhoneConnectionType | string;
	phoneType: EPhoneType | string;
	phone: string;
	phoneAreaCode: string;
	isPrincipal: boolean;
}

export interface IMultipleAddressItem {
	addressType: EAddressType | string;
	address: {
		principalStreet: string;
		latitude: number;
		longitude: number;
		streetNumber: string;
		provinceId: number;
		cantonId: number;
		parishId: number;
		isManualAddress: boolean;
	};
	isPrincipal: boolean;
	fullAddress: string;
	addressReference: string;
}

export interface IActionPanelOption {
	title: string | ReactNode;
	confirm?: {
		title: string;
		message: string | ReactNode;
	};
	route?: string;
	action?: () => void;
	hidden?: boolean;
	disabled?: boolean;
	target?: string;
}

export interface IPermissionBasic {
	id: number;
	name: string;
}
export interface IPermissionActions {
	update: number;
	delete: number;
	create: number;
	read: number;
	all_actions: number;
}

export interface IPermissionProgram extends IPermissionBasic {
	icon: string;
	actions: IPermissionActions;
	path: string;
	programs?: IPermissionProgram[];
}

export interface IPermissionSubmodule extends IPermissionBasic {
	icon: string;
	path: string;
	programs: IPermissionProgram[];
	groups?: IPermissionProgram[];
}
export interface IPermissionModule extends IPermissionBasic {
	submodules: IPermissionSubmodule[];
}

export interface IPermissionAgency extends IPermissionBasic {
	modules: IPermissionModule[];
}
export interface IPermission extends IPermissionBasic {
	agencies: IPermissionAgency[];
}

export interface IMenuItem {
	title: string;
	icon: string;
	id?: number;
	name?: string;
	path?: string;
	actions?: IPermissionActions;
	groups?: IMenuItem[];
	subList?: IMenuItem[];
}

export interface IUserInfo {
	identification: string;
	identificationType: string;
	name: string;
	picture: string;
	email: string;
	roles: {
		name: string;
	}[];
}

export interface IControllerProps {
	isForm?: boolean;
	isEditable?: boolean;
	isDetail?: boolean;
}

export interface IModalResponsiveProps {
	open: boolean;
	setOpen: (open: boolean) => void;
	onOk: () => void;
	setOnOk: (onOk: () => void) => void;
	onCancel: () => void;
	setOnCancel: (onCancel: () => void) => void;
	title: string;
	setTitle: (title: string) => void;
	footer: ReactNode;
	setFooter: (footer: ReactNode) => void;
	content: ReactNode;
	setContent: (content: ReactNode) => void;
}

export interface ITabsMaintenanceItem {
	key: string;
	label: string;
	children: ReactNode;
	icon?: ReactNode;
}
