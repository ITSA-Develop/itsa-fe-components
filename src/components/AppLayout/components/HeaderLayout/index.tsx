import { ActiveNotificationIcon, NotificationIcon, PinIcon, UserIcon } from '@/assets/icons';
import { DropdownCustomLabel } from '@/components/DropdownCustomLabel';
import { DropdownIcon } from '@/components/DropdownIcon';
import { Image } from '@/components/Image';
import { LOGO_DIMENSIONS, ROUTES_IMAGES } from '@/constants';
import { useSidebarStore } from '@/hooks';
import { ISelectOptionDropdownButton } from '@/interfaces';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { MenuUnfoldOutlined, SettingOutlined } from '@ant-design/icons';
import { Button, Dropdown, MenuProps } from 'antd';

export interface HeaderLayoutProps {
	loadingHeader: boolean;
	notifications?: MenuProps;
	userActions?: MenuProps;
	logo?: string;
}

export const HeaderLayout = ({
	loadingHeader,
	notifications = { items: [] },
	userActions = { items: [] },
	logo = '',
}: HeaderLayoutProps) => {
	const { setCurrentModule, setCurrentAgency } = useAppLayoutStore();
	const { collapsed, setCollapsed } = useSidebarStore();
	const { agencies } = useAppLayoutStore();
	const { currentAgency } = useAppLayoutStore();
	const { currentModule } = useAppLayoutStore();
	const { modulesAgency } = useAppLayoutStore();
	const { userName } = useAppLayoutStore();
	const { userRole } = useAppLayoutStore();

	const isActiveNotifications = Boolean(notifications?.items?.length);
	const isActiveUserActions = Boolean(userActions?.items?.length);

	const modulesData: ISelectOptionDropdownButton = {
		items: modulesAgency.map(m => ({
			key: m.id.toString(),
			label: m.name,
			onClick: (id: string) => handleSetCurrentModule(id),
		})),
	};

	const agenciesData: ISelectOptionDropdownButton = {
		items: agencies.map(a => ({
			key: a.id.toString(),
			label: a.name,
			onClick: (id: string) => handleSetCurrentAgency(id),
		})),
	};

	const handleSetCurrentModule = (moduleId: string) => {
		const module = modulesAgency.find(m => m.id.toString() === moduleId);
		if (module) {
			setCurrentModule(module);
		}
	};

	const handleSetCurrentAgency = (agencyId: string) => {
		const agency = agencies.find(a => a.id.toString() === agencyId);
		if (agency) {
			setCurrentAgency(agency);
		}
	};

	return (
		<header className="h-16">
			<div className="flex flex-row text-white-100 items-center justify-between w-full rounded-xl h-16 pr-4 pl-6 bg-primary-700">
				<div className="w-full flex flex-row items-center justify-start gap-4">
					{collapsed && (
						<Button
							type="text"
							icon={<MenuUnfoldOutlined className="text-white-100" />}
							onClick={() => setCollapsed(!collapsed)}
						/>
					)}
					<Image
						imgPath={logo || ROUTES_IMAGES.companyLogo}
						width={LOGO_DIMENSIONS.HEADER_WIDTH}
						height={LOGO_DIMENSIONS.HEADER_HEIGHT}
						alt="Logo"
						errorLabel="Importadora Tomebamba"
					/>
				</div>
				<div className="hidden sm:flex flex-row w-auto items-center gap-4">
					<div className="flex md:hidden flex-row items-center gap-4">
						<DropdownIcon
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentModule}
						/>
						<DropdownIcon
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentAgency}
						/>
					</div>

					<div className="hidden md:flex flex-row items-center gap-4">
						<DropdownCustomLabel
							defaultValue={currentModule?.id?.toString()}
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin mÃ³dulos asignados"
							onChange={handleSetCurrentModule}
						/>
						<DropdownCustomLabel
							defaultValue={currentAgency?.id?.toString()}
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin agencias asignadas"
							onChange={handleSetCurrentAgency}
						/>
						<div className="flex flex-col">
							<span className="text-4 whitespace-nowrap">{userName ?? ''}</span>
							<span className="text-primary-900 font-bold text-end text-xs whitespace-nowrap">{userRole ?? ''}</span>
						</div>
					</div>
				</div>
				<div className="flex flex-row pl-4">
					<Dropdown menu={userActions || { items: [] }} placement="bottomRight" disabled={!isActiveUserActions}>
						<Button type="text" icon={<UserIcon className="text-white-100 w-6 h-6" />} />
					</Dropdown>
					<Dropdown menu={notifications || { items: [] }} placement="bottomRight" disabled={!isActiveNotifications}>
						<Button
							type="text"
							icon={
								isActiveNotifications ? (
									<ActiveNotificationIcon className="text-white-100 w-6 h-6" />
								) : (
									<NotificationIcon className="text-white-100 w-6 h-6" />
								)
							}
						/>
					</Dropdown>
				</div>
			</div>
		</header>
	);
};
