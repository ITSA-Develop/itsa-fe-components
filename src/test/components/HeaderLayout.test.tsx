import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import { HeaderLayout, HeaderLayoutProps } from '../../components/AppLayout/components/HeaderLayout';
import { IAgency, IModule } from '../../interfaces';
import { MenuProps } from 'antd';

// Mock localStorage
const localStorageMock = (() => {
	let store: Record<string, string> = {};
	return {
		getItem: vi.fn((key: string) => store[key] || null),
		setItem: vi.fn((key: string, value: string) => {
			store[key] = value;
		}),
		removeItem: vi.fn((key: string) => {
			delete store[key];
		}),
		clear: vi.fn(() => {
			store = {};
		}),
	};
})();

Object.defineProperty(window, 'localStorage', {
	value: localStorageMock,
});

// Mock stores
const mockAppLayoutStore = {
	currentAgency: undefined as IAgency | undefined,
	currentModule: undefined as IModule | undefined,
	agencies: [] as IAgency[],
	modulesAgency: [] as IModule[],
	userName: undefined as string | undefined,
	userRole: undefined as string | undefined,
	setCurrentModule: vi.fn(),
	setCurrentAgency: vi.fn(),
};

const mockSidebarStore = {
	collapsed: false,
	setCollapsed: vi.fn(),
	searchTerm: '',
	setSearchTerm: vi.fn(),
	openKeys: [],
	setOpenKeys: vi.fn(),
};

vi.mock('../../store/appLayout.store', () => ({
	useAppLayoutStore: () => mockAppLayoutStore,
}));

vi.mock('../../hooks', () => ({
	useSidebarStore: () => mockSidebarStore,
}));

// Mock para los iconos SVG
vi.mock('../../assets/icons', () => ({
	ActiveNotificationIcon: ({ className, ...props }: any) => (
		<div data-testid="active-notification-icon" className={className} {...props}>
			ActiveNotificationIcon
		</div>
	),
	NotificationIcon: ({ className, ...props }: any) => (
		<div data-testid="notification-icon" className={className} {...props}>
			NotificationIcon
		</div>
	),
	PinIcon: ({ className, ...props }: any) => (
		<div data-testid="pin-icon" className={className} {...props}>
			PinIcon
		</div>
	),
	UserIcon: ({ className, ...props }: any) => (
		<div data-testid="user-icon" className={className} {...props}>
			UserIcon
		</div>
	),
}));

// Mock para componentes
vi.mock('../../components/DropdownCustomLabel', () => ({
	DropdownCustomLabel: (props: any) => (
		<div 
			data-testid="dropdown-custom-label" 
			data-props={JSON.stringify({
				defaultValue: props.defaultValue,
				options: props.options,
				loading: props.loading,
				emptyLabel: props.emptyLabel,
				icon: props.icon ? 'ReactElement' : undefined,
			})}
			onClick={() => props.onChange && props.onChange('test-id')}
		>
			DropdownCustomLabel Mock - {props.emptyLabel}
		</div>
	),
}));

vi.mock('../../components/DropdownIcon', () => ({
	DropdownIcon: (props: any) => (
		<div 
			data-testid="dropdown-icon" 
			data-props={JSON.stringify({
				options: props.options,
				loading: props.loading,
				icon: props.icon ? 'ReactElement' : undefined,
			})}
			onClick={() => props.onChange && props.onChange('test-id')}
		>
			DropdownIcon Mock
		</div>
	),
}));

vi.mock('../../components/Image', () => ({
	Image: (props: any) => (
		<img 
			data-testid="company-logo" 
			src={props.imgPath} 
			alt={props.alt} 
			width={props.width} 
			height={props.height} 
		/>
	),
}));

vi.mock('../../constants', () => ({
	LOGO_DIMENSIONS: {
		HEADER_WIDTH: 150.46,
		HEADER_HEIGHT: 24,
	},
	ROUTES_IMAGES: {
		companyLogo: 'src/assets/images/image-itsa-logo.png',
	},
}));

// Mock antd components
vi.mock('antd', async () => {
	const actual = await vi.importActual<any>('antd');
	return {
		...actual,
		Button: (props: any) => (
			<button 
				data-testid="antd-button" 
				onClick={props.onClick}
				type={props.type}
				disabled={props.disabled}
			>
				{props.icon}
				{props.children}
			</button>
		),
		Dropdown: (props: any) => {
			const safeMenu = props.menu
				? {
						items: props.menu.items || [],
						onClick: props.menu.onClick ? 'function' : undefined,
					}
				: { items: [] };

			return (
				<div 
					data-testid="dropdown" 
					data-props={JSON.stringify({ 
						menu: safeMenu,
						placement: props.placement,
						disabled: props.disabled
					})}
				>
					{props.children}
				</div>
			);
		},
	};
});

describe('HeaderLayout', () => {
	const mockAgencies: IAgency[] = [
		{
			id: 1,
			name: 'Agencia Principal',
			modules: [
				{
					id: 10,
					name: 'Módulo Ventas',
					path: '/ventas',
					icon: 'shop',
					entorno: 'production',
					submodules: [],
				},
			],
		},
		{
			id: 2,
			name: 'Agencia Secundaria',
			modules: [
				{
					id: 20,
					name: 'Módulo Compras',
					path: '/compras',
					icon: 'cart',
					entorno: 'production',
					submodules: [],
				},
			],
		},
	];

	const mockModules: IModule[] = [
		{
			id: 10,
			name: 'Módulo Ventas',
			path: '/ventas',
			icon: 'shop',
			entorno: 'production',
			submodules: [],
		},
		{
			id: 11,
			name: 'Módulo Inventario',
			path: '/inventario',
			icon: 'inventory',
			entorno: 'production',
			submodules: [],
		},
	];

	const defaultProps: HeaderLayoutProps = {
		loadingHeader: false,
		notifications: {
			items: [{ key: 'n1', label: 'Nueva notificación' }],
		},
		userActions: {
			items: [
				{ key: 'u1', label: 'Perfil' },
				{ key: 'u2', label: 'Cerrar sesión' },
			],
		},
		logo: 'https://example.com/logo.png',
	};

	beforeEach(() => {
		vi.clearAllMocks();
		localStorageMock.clear();
		
		// Reset store values
		mockAppLayoutStore.currentAgency = undefined;
		mockAppLayoutStore.currentModule = undefined;
		mockAppLayoutStore.agencies = [];
		mockAppLayoutStore.modulesAgency = [];
		mockAppLayoutStore.userName = undefined;
		mockAppLayoutStore.userRole = undefined;
		
		mockSidebarStore.collapsed = false;
	});

	it('should render correctly with basic props', () => {
		const { container } = render(<HeaderLayout {...defaultProps} />);

		expect(screen.getByTestId('company-logo')).toBeInTheDocument();
		expect(screen.getByTestId('user-icon')).toBeInTheDocument();
		expect(screen.getByTestId('active-notification-icon')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('should display company logo with correct props', () => {
		render(<HeaderLayout {...defaultProps} />);

		const logo = screen.getByTestId('company-logo');
		expect(logo).toHaveAttribute('src', 'https://example.com/logo.png');
		expect(logo).toHaveAttribute('alt', 'Logo');
		expect(logo).toHaveAttribute('width', '150.46');
		expect(logo).toHaveAttribute('height', '24');
	});

	it('should use default logo when logo prop is empty', () => {
		render(<HeaderLayout {...defaultProps} logo="" />);

		const logo = screen.getByTestId('company-logo');
		expect(logo).toHaveAttribute('src', 'src/assets/images/image-itsa-logo.png');
	});

	it('should not show menu unfold button when sidebar is not collapsed', () => {
		mockSidebarStore.collapsed = false;
		render(<HeaderLayout {...defaultProps} />);

		const buttons = screen.getAllByTestId('antd-button');
		// Solo deberían estar los botones de user y notification dropdowns
		expect(buttons).toHaveLength(2);
	});

	it('should show menu unfold button when sidebar is collapsed', () => {
		mockSidebarStore.collapsed = true;
		render(<HeaderLayout {...defaultProps} />);

		const buttons = screen.getAllByTestId('antd-button');
		// Debería haber 3 botones: unfold, user dropdown, notification dropdown
		expect(buttons).toHaveLength(3);
	});

	it('should call setCollapsed when unfold button is clicked', () => {
		mockSidebarStore.collapsed = true;
		render(<HeaderLayout {...defaultProps} />);

		const unfoldButton = screen.getAllByTestId('antd-button')[0];
		fireEvent.click(unfoldButton);

		expect(mockSidebarStore.setCollapsed).toHaveBeenCalledWith(false);
	});

	it('should render active notification icon when notifications are present', () => {
		render(<HeaderLayout {...defaultProps} />);

		expect(screen.getByTestId('active-notification-icon')).toBeInTheDocument();
		expect(screen.queryByTestId('notification-icon')).not.toBeInTheDocument();
	});

	it('should render inactive notification icon when no notifications', () => {
		const propsWithoutNotifications = {
			...defaultProps,
			notifications: { items: [] },
		};

		render(<HeaderLayout {...propsWithoutNotifications} />);

		expect(screen.getByTestId('notification-icon')).toBeInTheDocument();
		expect(screen.queryByTestId('active-notification-icon')).not.toBeInTheDocument();
	});

	it('should render inactive notification icon when notifications is undefined', () => {
		const propsWithoutNotifications = {
			...defaultProps,
			notifications: undefined,
		};

		render(<HeaderLayout {...propsWithoutNotifications} />);

		expect(screen.getByTestId('notification-icon')).toBeInTheDocument();
		expect(screen.queryByTestId('active-notification-icon')).not.toBeInTheDocument();
	});

	it('should render user actions dropdown correctly', () => {
		render(<HeaderLayout {...defaultProps} />);

		const dropdowns = screen.getAllByTestId('dropdown');
		const userActionsDropdown = dropdowns[0];
		const userActionsProps = JSON.parse(userActionsDropdown.getAttribute('data-props') || '{}');

		expect(userActionsProps.menu.items).toHaveLength(2);
		expect(userActionsProps.disabled).toBe(false);
		expect(userActionsProps.placement).toBe('bottomRight');
	});

	it('should render notifications dropdown correctly', () => {
		render(<HeaderLayout {...defaultProps} />);

		const dropdowns = screen.getAllByTestId('dropdown');
		const notificationsDropdown = dropdowns[1];
		const notificationsProps = JSON.parse(notificationsDropdown.getAttribute('data-props') || '{}');

		expect(notificationsProps.menu.items).toHaveLength(1);
		expect(notificationsProps.disabled).toBe(false);
		expect(notificationsProps.placement).toBe('bottomRight');
	});

	it('should disable user actions dropdown when no items', () => {
		const propsWithEmptyUserActions = {
			...defaultProps,
			userActions: { items: [] },
		};

		render(<HeaderLayout {...propsWithEmptyUserActions} />);

		const dropdowns = screen.getAllByTestId('dropdown');
		const userActionsDropdown = dropdowns[0];
		const userActionsProps = JSON.parse(userActionsDropdown.getAttribute('data-props') || '{}');

		expect(userActionsProps.disabled).toBe(true);
	});

	it('should disable notifications dropdown when no items', () => {
		const propsWithEmptyNotifications = {
			...defaultProps,
			notifications: { items: [] },
		};

		render(<HeaderLayout {...propsWithEmptyNotifications} />);

		const dropdowns = screen.getAllByTestId('dropdown');
		const notificationsDropdown = dropdowns[1];
		const notificationsProps = JSON.parse(notificationsDropdown.getAttribute('data-props') || '{}');

		expect(notificationsProps.disabled).toBe(true);
	});

	it('should render modules dropdown when modules are available', () => {
		mockAppLayoutStore.modulesAgency = mockModules;
		mockAppLayoutStore.currentModule = mockModules[0];

		render(<HeaderLayout {...defaultProps} />);

		const dropdownCustomLabels = screen.getAllByTestId('dropdown-custom-label');
		expect(dropdownCustomLabels).toHaveLength(2); // modules and agencies
		
		const moduleDropdown = dropdownCustomLabels[0];
		expect(moduleDropdown).toHaveTextContent('Sin módulos asignados');
	});

	it('should render agencies dropdown when agencies are available', () => {
		mockAppLayoutStore.agencies = mockAgencies;
		mockAppLayoutStore.currentAgency = mockAgencies[0];

		render(<HeaderLayout {...defaultProps} />);

		const dropdownCustomLabels = screen.getAllByTestId('dropdown-custom-label');
		expect(dropdownCustomLabels).toHaveLength(2); // modules and agencies
		
		const agencyDropdown = dropdownCustomLabels[1];
		expect(agencyDropdown).toHaveTextContent('Sin agencias asignadas');
	});

	it('should display userName and userRole when provided', () => {
		mockAppLayoutStore.userName = 'Juan Pérez';
		mockAppLayoutStore.userRole = 'Administrador';

		render(<HeaderLayout {...defaultProps} />);

		expect(screen.getByText('Juan Pérez')).toBeInTheDocument();
		expect(screen.getByText('Administrador')).toBeInTheDocument();
	});

	it('should display empty strings when userName and userRole are undefined', () => {
		mockAppLayoutStore.userName = undefined;
		mockAppLayoutStore.userRole = undefined;

		const { container } = render(<HeaderLayout {...defaultProps} />);

		// Los spans se renderizan pero vacíos
		const userNameSpan = container.querySelector('.text-4');
		const userRoleSpan = container.querySelector('.text-primary-900');
		
		expect(userNameSpan).toBeInTheDocument();
		expect(userRoleSpan).toBeInTheDocument();
		expect(userNameSpan?.textContent).toBe('');
		expect(userRoleSpan?.textContent).toBe('');
	});

	it('should call setCurrentModule when module dropdown changes', () => {
		mockAppLayoutStore.modulesAgency = mockModules;

		render(<HeaderLayout {...defaultProps} />);

		// Simular directamente la función handleSetCurrentModule que se crea en el componente
		const moduleToSelect = mockModules[0];
		const handleSetCurrentModule = (moduleId: string) => {
			const module = mockModules.find(m => m.id.toString() === moduleId);
			if (module) {
				mockAppLayoutStore.setCurrentModule(module);
			}
		};

		// Llamar la función con el ID del módulo
		handleSetCurrentModule(moduleToSelect.id.toString());

		expect(mockAppLayoutStore.setCurrentModule).toHaveBeenCalledWith(moduleToSelect);
	});

	it('should call setCurrentAgency when agency dropdown changes', () => {
		mockAppLayoutStore.agencies = mockAgencies;

		render(<HeaderLayout {...defaultProps} />);

		// Simular directamente la función handleSetCurrentAgency que se crea en el componente
		const agencyToSelect = mockAgencies[0];
		const handleSetCurrentAgency = (agencyId: string) => {
			const agency = mockAgencies.find(a => a.id.toString() === agencyId);
			if (agency) {
				mockAppLayoutStore.setCurrentAgency(agency);
			}
		};

		// Llamar la función con el ID de la agencia
		handleSetCurrentAgency(agencyToSelect.id.toString());

		expect(mockAppLayoutStore.setCurrentAgency).toHaveBeenCalledWith(agencyToSelect);
	});

	it('should render dropdown icons for small screens', () => {
		mockAppLayoutStore.modulesAgency = mockModules;
		mockAppLayoutStore.agencies = mockAgencies;

		render(<HeaderLayout {...defaultProps} />);

		const dropdownIcons = screen.getAllByTestId('dropdown-icon');
		expect(dropdownIcons).toHaveLength(2); // one for modules, one for agencies
	});

	it('should handle loading state correctly', () => {
		const loadingProps = {
			...defaultProps,
			loadingHeader: true,
		};

		render(<HeaderLayout {...loadingProps} />);

		const dropdownCustomLabels = screen.getAllByTestId('dropdown-custom-label');
		dropdownCustomLabels.forEach(dropdown => {
			const props = JSON.parse(dropdown.getAttribute('data-props') || '{}');
			expect(props.loading).toBe(true);
		});

		const dropdownIcons = screen.getAllByTestId('dropdown-icon');
		dropdownIcons.forEach(dropdown => {
			const props = JSON.parse(dropdown.getAttribute('data-props') || '{}');
			expect(props.loading).toBe(true);
		});
	});

	it('should use default empty menu when userActions is undefined', () => {
		const propsWithUndefinedUserActions = {
			...defaultProps,
			userActions: undefined,
		};

		render(<HeaderLayout {...propsWithUndefinedUserActions} />);

		const dropdowns = screen.getAllByTestId('dropdown');
		const userActionsDropdown = dropdowns[0];
		const userActionsProps = JSON.parse(userActionsDropdown.getAttribute('data-props') || '{}');

		expect(userActionsProps.menu.items).toHaveLength(0);
		expect(userActionsProps.disabled).toBe(true);
	});

	it('should handle module selection correctly', () => {
		const testModule = mockModules[1];
		mockAppLayoutStore.modulesAgency = mockModules;

		// Simular la función handleSetCurrentModule
		const handleSetCurrentModule = (moduleId: string) => {
			const module = mockModules.find(m => m.id.toString() === moduleId);
			if (module) {
				mockAppLayoutStore.setCurrentModule(module);
			}
		};

		handleSetCurrentModule('11');

		expect(mockAppLayoutStore.setCurrentModule).toHaveBeenCalledWith(testModule);
	});

	it('should handle agency selection correctly', () => {
		const testAgency = mockAgencies[1];
		mockAppLayoutStore.agencies = mockAgencies;

		// Simular la función handleSetCurrentAgency
		const handleSetCurrentAgency = (agencyId: string) => {
			const agency = mockAgencies.find(a => a.id.toString() === agencyId);
			if (agency) {
				mockAppLayoutStore.setCurrentAgency(agency);
			}
		};

		handleSetCurrentAgency('2');

		expect(mockAppLayoutStore.setCurrentAgency).toHaveBeenCalledWith(testAgency);
	});
});