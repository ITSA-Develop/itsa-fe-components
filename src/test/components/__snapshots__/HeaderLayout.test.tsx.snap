// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderLayout > should render correctly with basic props 1`] = `
<div>
  <header
    class="h-16"
  >
    <div
      class="flex flex-row text-white-100 items-center justify-between w-full rounded-xl h-16 pr-4 pl-6 bg-primary-700"
    >
      <div
        class="w-full flex flex-row items-center justify-start gap-4"
      >
        <img
          alt="Logo"
          data-testid="company-logo"
          height="24"
          src="https://example.com/logo.png"
          width="150.46"
        />
      </div>
      <div
        class="hidden sm:flex flex-row w-auto items-center gap-4"
      >
        <div
          class="flex md:hidden flex-row items-center gap-4"
        >
          <div
            data-props="{"options":{"items":[]},"loading":false,"icon":"ReactElement"}"
            data-testid="dropdown-icon"
          >
            DropdownIcon Mock
          </div>
          <div
            data-props="{"options":{"items":[]},"loading":false,"icon":"ReactElement"}"
            data-testid="dropdown-icon"
          >
            DropdownIcon Mock
          </div>
        </div>
        <div
          class="hidden md:flex flex-row items-center gap-4"
        >
          <div
            data-props="{"options":{"items":[]},"loading":false,"emptyLabel":"Sin m贸dulos asignados","icon":"ReactElement"}"
            data-testid="dropdown-custom-label"
          >
            DropdownCustomLabel Mock - 
            Sin m贸dulos asignados
          </div>
          <div
            data-props="{"options":{"items":[]},"loading":false,"emptyLabel":"Sin agencias asignadas","icon":"ReactElement"}"
            data-testid="dropdown-custom-label"
          >
            DropdownCustomLabel Mock - 
            Sin agencias asignadas
          </div>
          <div
            class="flex flex-col"
          >
            <span
              class="text-4 whitespace-nowrap"
            />
            <span
              class="text-primary-900 font-bold text-end text-xs whitespace-nowrap"
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-row pl-4"
      >
        <div
          data-props="{"menu":{"items":[{"key":"u1","label":"Perfil"},{"key":"u2","label":"Cerrar sesi贸n"}]},"placement":"bottomRight","disabled":false}"
          data-testid="dropdown"
        >
          <button
            data-testid="antd-button"
            type="text"
          >
            <div
              class="text-white-100 w-6 h-6"
              data-testid="user-icon"
            >
              UserIcon
            </div>
          </button>
        </div>
        <div
          data-props="{"menu":{"items":[{"key":"n1","label":"Nueva notificaci贸n"}]},"placement":"bottomRight","disabled":false}"
          data-testid="dropdown"
        >
          <button
            data-testid="antd-button"
            type="text"
          >
            <div
              class="text-white-100 w-6 h-6"
              data-testid="active-notification-icon"
            >
              ActiveNotificationIcon
            </div>
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
